clients/pkg/promtail/targets/gcplog/push_translation.go

```diff
@@ -37,81 +27,61 @@ type PushMessage struct {
+	}
+	return nil
+}
+
+// translate converts a GCP PushMessage into a loki api.Entry. It parses the push-specific labels, and delegates the rest to parseGCPLogsEntry.
 func translate(m PushMessage, other model.LabelSet, useIncomingTimestamp bool, relabelConfigs []*relabel.Config, xScopeOrgID string) (api.Entry, error) {
-	// mandatory label for gcplog
+	// Collect all push-specific labels. Every one of them is first configured as optional, and the user
+	// can relabel it if needed. The relabeling and internal drop is handled in parseGCPLogsEntry.
 	lbs := labels.NewBuilder(nil)
 	lbs.Set("__gcp_message_id", m.Message.ID)
-
-	// labels from gcp log entry. Add it as internal labels
+	lbs.Set("__gcp_subscription_name", m.Subscription)
 	for k, v := range m.Message.Attributes {
 		lbs.Set(fmt.Sprintf("__gcp_attributes_%s", convertToLokiCompatibleLabel(k)), v)
 	}
 
-	var processed labels.Labels
+	// Add fixed labels coming from the target configuration
+	fixedLabels := other.Clone()
 
-	// apply relabeling
-	if len(relabelConfigs) > 0 {
-		processed = relabel.Process(lbs.Labels(), relabelConfigs...)
-	} else {
-		processed = lbs.Labels()
-	}
-
-	// final labelset that will be sent to loki
-	labels := make(model.LabelSet)
-	for _, lbl := range processed {
-		// ignore internal labels
-		if strings.HasPrefix(lbl.Name, "__") {
-			continue
-		}
-		// ignore invalid labels
-		if !model.LabelName(lbl.Name).IsValid() || !model.LabelValue(lbl.Value).IsValid() {
-			continue
-		}
-		labels[model.LabelName(lbl.Name)] = model.LabelValue(lbl.Value)
+	// If the incoming request carries the tenant id, inject it as the reserved label, so it's used by the
+	// remote write client.
+	if xScopeOrgID != "" {
+		fixedLabels[lokiClient.ReservedLabelTenantID] = model.LabelValue(xScopeOrgID)
 	}
 
-	// add labels coming from scrapeconfig
-	labels = labels.Merge(other)
-
-	ts := time.Now()
-
 	decodedData, err := base64.StdEncoding.DecodeString(m.Message.Data)
 	if err != nil {
 		return api.Entry{}, fmt.Errorf("failed to decode data: %w", err)
 	}
-	line := string(decodedData)
 
-	if useIncomingTimestamp {
-		var err error
-		ts, err = time.Parse(time.RFC3339, m.Message.PublishTimestamp)
-		if err != nil {
-			return api.Entry{}, fmt.Errorf("invalid publish timestamp format: %w", err)
-		}
+	entry, err := parseGCPLogsEntry(decodedData, fixedLabels, lbs.Labels(), useIncomingTimestamp, relabelConfigs)
+	if err != nil {
+		return api.Entry{}, fmt.Errorf("failed to parse logs entry: %w", err)
 	}
 
-	// If the incoming request carries the tenant id, inject it as the reserved label, so it's used by the
-	// remote write client.
-	if xScopeOrgID != "" {
-		labels[lokiClient.ReservedLabelTenantID] = model.LabelValue(xScopeOrgID)
-	}
+	return entry, nil
+}
+
+var separatorCharacterReplacer *strings.Replacer
 
-	return api.Entry{
-		Labels: labels,
-		Entry: logproto.Entry{
-			Timestamp: ts,
-			Line:      line,
-		},
-	}, nil
+func init() {
```

Comment:
Can you please move it to global const? (I guess no reason to use `init()` here?)